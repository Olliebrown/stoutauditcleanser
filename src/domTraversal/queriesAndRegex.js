// Various queries and regex strings used throughout
export const QUERIES = {
  programHeader: '[id*="DERIVED_SAA_DPR_GROUPBOX1GP"]',
  requirementHeader: '.PAGROUPDIVIDER',
  expander: '#DERIVED_SAA_DPR_SSS_EXPAND_ALL',
  expanderList: 'a[id^="DERIVED_SAA_DPR_GROUPBOX1"][aria-expanded="false"]',
  programTable: (depth) => `tr:nth-child(2) > td tbody > tr:nth-child(2) > td:nth-child(2) tbody > tr:nth-child(2) td tbody > tr:nth-child(${depth}) > td:nth-child(2) tbody tbody`
}

export const REGEX = {
  generalHeader: /(?:GENERAL INFORMATION)/i,
  universityRequirements: /(?:UNIVERSITY REQUIREMENTS)|(?:HONORS COLLEGE OVERVIEW)/i,
  requirementDescription: /\s*(?<satisfied>(?:Not Satisfied)|(?:Satisfied)):\s*(?<ID>[\w-]+):\s*(?<description>.*)\s*/i,
  informationalOnly: /(?:Purpose of Academic Advisement Report)|(?:In-Progress Repeat Coursework)|(?:REMEDIAL\/PLACEMENT COURSEWORK INFORMATION)|(?:Graduation Information)|(?:Graduation With Honors Policy)/i
}
