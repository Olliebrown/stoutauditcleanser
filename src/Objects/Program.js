import { QUERIES, REGEX } from '../domTraversal/queriesAndRegex.js'

import AuditNode from './AuditNode.js'
import { makeRequirementsArray, makeSubRequirementsArray } from './NodeFactories.js'

/**
 * An object for examining a program or sub-program and its requirements
 */
export default class Program extends AuditNode {
  // References to DOM elements used for extraction of details
  #programHeaderNode = null
  #programRoot = null
  #mainTable = null

  constructor (programHeaderNode) {
    // Required call to parent constructor
    super()

    // Initialize important members used for extraction below
    this.#programHeaderNode = programHeaderNode
    this.#programRoot = programHeaderNode.parentNode.parentNode.parentNode

    // Try to find the main program table
    this.#mainTable = this.#programRoot.querySelector(QUERIES.programTable(5))
    if (!this.#mainTable) {
      // Try reduced depth
      this.#mainTable = this.#programRoot.querySelector(QUERIES.programTable(3))
      if (!this.#mainTable) {
        throw new Error('Failed to find program table')
      }
    }

    // Initialize internal values
    this._initialize(this.#programRoot, /RG-\d+/)
  }

  /**
   * Extract and return just the text of the program's header
   * @returns {string} The text within the program's header row
   * @override
   */
  _extractHeading () {
    return this.#programHeaderNode.textContent
  }

  // Find and extract the completion unit information
  _extractFullText () {
    const descriptionText = this.#programRoot.querySelector(QUERIES.programTable(3))?.textContent
    if (descriptionText?.match(REGEX.units)) {
      return descriptionText
    }

    return super._extractFullText()
  }

  /**
   * Extract the requirements and convert them to an array of Requirement objects
   * @returns {Array(Requirement)} Array of Requirement objects extracted from the HTML
   * @override
   */
  _extractSubNodes () {
    // Try to extract as Requirements
    let requirements = makeRequirementsArray(this.#mainTable)
    if (!Array.isArray(requirements) || requirements.length < 1) {
      // Try sub-requirements instead
      requirements = makeSubRequirementsArray(this.#mainTable)
    }
    return requirements
  }

  /**
   * If all requirements are satisfied, then this is considered satisfied
   * @returns {bool} Weather or not this program is satisfied (e.g. completed)
   * @override
   */
  isSatisfied () {
    const requirements = this.getSubNodes()

    // Something is wrong so just return incomplete
    if (!Array.isArray(requirements) || requirements.length < 1) {
      return AuditNode.SATISFIED_TYPE.INCOMPLETE
    }

    // Check each requirement
    return requirements.reduce((complete, requirement) => {
      switch (complete) {
        // Once we have an incomplete, we can't go back
        case AuditNode.SATISFIED_TYPE.INCOMPLETE:
          return AuditNode.SATISFIED_TYPE.INCOMPLETE

        // Never return to 'complete' status once we are in-progress
        case AuditNode.SATISFIED_TYPE.IN_PROGRESS:
          if (requirement.isSatisfied() === AuditNode.SATISFIED_TYPE.COMPLETE) {
            return AuditNode.SATISFIED_TYPE.IN_PROGRESS
          }
      }

      // Otherwise, just return the status of the current requirement
      return requirement.isSatisfied()
    }, AuditNode.SATISFIED_TYPE.COMPLETE)
  }
}
