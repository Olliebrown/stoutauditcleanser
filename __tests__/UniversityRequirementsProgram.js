// Necessary when using ECMA modules
import { jest } from '@jest/globals'

// Setup basic HTML page before each test
import fs from 'fs'
import path from 'path'

// Project specific imports
import { QUERIES } from '../src/domTraversal/queriesAndRegex.js'
import Program from '../src/Objects/Program.js'

// Read in the HTML page
const baseHTML = fs.readFileSync(path.join('__tests__', 'data', 'UniversityProgram.html'), 'utf8')

// Disable jest mocking for fs module
jest.dontMock('fs')

describe('General Info Node', () => {
  // Inject clean version of HTML page and rebuild program before tests
  let program = null
  beforeAll(() => {
    // Load page from imported HTML
    document.documentElement.innerHTML = baseHTML.toString()

    // Extract program headers and make sure there's at least one
    const headerNodes = document.querySelectorAll(QUERIES.programHeader)
    if (headerNodes?.length < 1) {
      throw new Error('Unable to find at least one program')
    }

    // Create program object
    program = new Program(headerNodes[0])
  })

  // Clear out the HTML and program
  afterAll(() => {
    jest.resetModules()
    program = null
  })

  // Basic page structure tests
  test('should have a specific page title and specific program name', () => {
    // Check title
    const title = document.querySelector('title')
    expect(title).toBeTruthy()
    expect(title.textContent).toBe('University Requirements')

    // Check program name
    expect(program).toBeTruthy()
    expect(program.getName()).toBe('University Requirements For Graduation')
  })
})
