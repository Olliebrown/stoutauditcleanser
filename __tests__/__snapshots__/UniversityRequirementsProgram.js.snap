// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`University Requirements Node should have expected snapshot 1`] = `
{
  "internalId": "RG-0001",
  "key": "university_requirements_for_graduation",
  "name": "University Requirements For Graduation",
  "satisfied": "INCOMPLETE",
  "subNodes": [
    {
      "description": "UNIVERSITY CREDIT REQUIREMENT FOR GRADUATION -  120 Earned Credits Required (Some Academic Plans may require additional credits for completion of degree requirements. Contact your Program Director and/or Academic Advisor for further information.)",
      "index": 1,
      "internalId": "RQ-10029",
      "isGenEd": false,
      "key": "university_credit_requirement_for_graduation",
      "name": "University Credit Requirement For Graduation",
      "nextIndex": 5,
      "satisfied": "INCOMPLETE",
      "subNodes": [],
      "units": {
        "credits": {
          "current": 111,
          "required": 120,
        },
      },
    },
    {
      "description": "UNIVERSITY GPA REQUIREMENT FOR GRADUATION - An 2.00 Overall GPA is Required (Some Academic Plans may require a higher GPA for completion of degree requirements.  Contact your Program Director and/or Academic Advisor for further information.) ",
      "index": 5,
      "internalId": "RQ-0183",
      "isGenEd": false,
      "key": "university_gpa_requirement_for_graduation",
      "name": "University GPA Requirement For Graduation",
      "nextIndex": 9,
      "satisfied": "COMPLETE",
      "subNodes": [],
      "units": {
        "gpa": {
          "current": 2.82,
          "required": 2,
        },
      },
    },
    {
      "description": "UNIVERSITY RESIDENCY REQUIREMENT FOR GRADUATION: UNIVERSITY RESIDENCY REQUIREMENT FOR GRADUATION -",
      "index": 9,
      "internalId": "RQ-10136",
      "isGenEd": false,
      "key": "university_residency_requirement_for_graduation",
      "name": "University Residency Requirement For Graduation",
      "nextIndex": 12,
      "satisfied": "COMPLETE",
      "subNodes": [],
      "units": {
        "credits": {
          "current": 32,
          "required": 32,
        },
      },
    },
  ],
  "units": {
    "credits": {
      "current": 111,
      "required": 120,
    },
    "gpa": {
      "current": 2.82,
      "required": 2,
    },
  },
}
`;
